File: server/src/Firewall.js
const {log} = Candy.core('Log', false).init('Firewall')

/**
 * Firewall class to handle IP blocking and rate limiting for all services (Web, DNS, Mail, etc).
 */
class Firewall {
  #requestCounts = new Map() // IP -> { count, timestamp }
  #config = {}
  #cleanupInterval = null

  constructor() {
    this.load()
    // Run cleanup every minute
    this.#cleanupInterval = setInterval(() => this.cleanup(), 60000)
    if (this.#cleanupInterval.unref) this.#cleanupInterval.unref()
  }

  /**
   * Load configuration from the global Config module.
   */
  load() {
    // Load configuration from Candy.core('Config').config
    const config = Candy.core('Config').config.firewall || {}

    this.#config = {
      enabled: config.enabled !== false,
      rateLimit: {
        enabled: config.rateLimit?.enabled !== false,
        windowMs: config.rateLimit?.windowMs ?? 60000, // 1 minute
        max: config.rateLimit?.max ?? 300 // limit each IP to 300 requests per windowMs
      },
      blacklist: new Set(config.blacklist || []),
      whitelist: new Set(config.whitelist || [])
    }
  }

  /**
   * Check if an IP should be allowed.
   * @param {string} ip - The IP address to check.
   * @returns {Object} An object containing {allowed: boolean, reason?: string}.
   */
  check(ip) {
    if (!this.#config.enabled) return {allowed: true}

    // Normalize IPv6-mapped IPv4 addresses
    if (ip && ip.startsWith('::ffff:')) {
      ip = ip.substring(7)
    }

    if (!ip) return {allowed: true}

    // 1. Check whitelist (bypass everything)
    if (this.#config.whitelist.has(ip)) return {allowed: true}

    // 2. Check blacklist
    if (this.#config.blacklist.has(ip)) {
      log(`Blocked request from blacklisted IP: ${ip}`)
      return {allowed: false, reason: 'blacklist'}
    }

    // 3. Rate limiting
    if (this.#config.rateLimit.enabled) {
      // Memory protection: if map gets too big, clear it to prevent memory leak
      if (this.#requestCounts.size > 20000) {
        this.#requestCounts.clear()
        log('Firewall request counts cleared due to memory limit')
      }

      const now = Date.now()
      let record = this.#requestCounts.get(ip)

      if (!record) {
        record = {count: 0, timestamp: now}
        this.#requestCounts.set(ip, record)
      }

      // Check window
      if (now - record.timestamp > this.#config.rateLimit.windowMs) {
        // Reset window
        record.count = 1
        record.timestamp = now
      } else {
        record.count++
      }

      if (record.count > this.#config.rateLimit.max) {
        if (record.count === this.#config.rateLimit.max + 1) {
          log(`Rate limit exceeded for IP: ${ip}`)
        }
        return {allowed: false, reason: 'rate_limit'}
      }
    }

    return {allowed: true}
  }

  /**
   * Cleanup stale rate limit records.
   */
  cleanup() {
    if (!this.#config.rateLimit?.windowMs) return

    const now = Date.now()
    const windowMs = this.#config.rateLimit.windowMs

    for (const [ip, record] of this.#requestCounts) {
      if (now - record.timestamp > windowMs) {
        this.#requestCounts.delete(ip)
      }
    }
  }

  /**
   * Add an IP to the blacklist.
   * @param {string} ip - The IP address to block.
   */
  addBlock(ip) {
    if (this.#config.whitelist.has(ip)) this.#config.whitelist.delete(ip)
    this.#config.blacklist.add(ip)
    this.#save()
  }

  /**
   * Remove an IP from the blacklist.
   * @param {string} ip - The IP address to unblock.
   */
  removeBlock(ip) {
    this.#config.blacklist.delete(ip)
    this.#save()
  }

  /**
   * Add an IP to the whitelist.
   * @param {string} ip - The IP address to whitelist.
   */
  addWhitelist(ip) {
    if (this.#config.blacklist.has(ip)) this.#config.blacklist.delete(ip)
    this.#config.whitelist.add(ip)
    this.#save()
  }

  /**
   * Remove an IP from the whitelist.
   * @param {string} ip - The IP address to remove from whitelist.
   */
  removeWhitelist(ip) {
    this.#config.whitelist.delete(ip)
    this.#save()
  }

  #save() {
    // Update the global config
    if (!Candy.core('Config').config.firewall) Candy.core('Config').config.firewall = {}

    Candy.core('Config').config.firewall.blacklist = Array.from(this.#config.blacklist)
    Candy.core('Config').config.firewall.whitelist = Array.from(this.#config.whitelist)
  }
}

module.exports = Firewall

File: server/proxy/proxy/firewall.go
package proxy

import (
	"log"
	"net"
	"net/http"
	"strings"
	"sync"
	"time"

	"odac-proxy/config"
)

type Firewall struct {
	config        config.Firewall
	blacklistMap  map[string]struct{}
	whitelistMap  map[string]struct{}
	requestCounts map[string]*requestRecord
	wsCounts      map[string]int // Active WebSocket connections per IP
	mu            sync.RWMutex
	stopCleanup   chan struct{}
}

type requestRecord struct {
	count     int
	timestamp int64
}

func NewFirewall(cfg config.Firewall) *Firewall {
	f := &Firewall{
		config:        cfg,
		blacklistMap:  sliceToMap(cfg.Blacklist),
		whitelistMap:  sliceToMap(cfg.Whitelist),
		requestCounts: make(map[string]*requestRecord),
		wsCounts:      make(map[string]int),
		stopCleanup:   make(chan struct{}),
	}
	go f.startCleanupLoop()
	return f
}

func (f *Firewall) UpdateConfig(cfg config.Firewall) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.config = cfg
	f.blacklistMap = sliceToMap(cfg.Blacklist)
	f.whitelistMap = sliceToMap(cfg.Whitelist)
}

func (f *Firewall) GetRequestTimeout() int {
	f.mu.RLock()
	defer f.mu.RUnlock()
	return f.config.RequestTimeout
}

func (f *Firewall) startCleanupLoop() {
	ticker := time.NewTicker(1 * time.Minute)
	for {
		select {
		case <-ticker.C:
			f.cleanup()
		case <-f.stopCleanup:
			ticker.Stop()
			return
		}
	}
}

func (f *Firewall) cleanup() {
	f.mu.Lock()
	defer f.mu.Unlock()

	if f.config.RateLimit.WindowMs == 0 {
		return
	}

	now := time.Now().UnixMilli()
	windowMs := int64(f.config.RateLimit.WindowMs)

	for ip, record := range f.requestCounts {
		if now-record.timestamp > windowMs {
			delete(f.requestCounts, ip)
		}
	}
}

func (f *Firewall) Check(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		f.mu.RLock()
		if !f.config.Enabled {
			f.mu.RUnlock()
			next.ServeHTTP(w, r)
			return
		}
		
		// Copy config values needed for checking to avoid holding RLock too long if possible,
		// but checking slice contains is fast enough to keep lock.
		// However, we need to upgrade lock for rate limiting.
		
		blacklistMap := f.blacklistMap
		whitelistMap := f.whitelistMap
		rateLimit := f.config.RateLimit
		f.mu.RUnlock()

		ip, _, err := net.SplitHostPort(r.RemoteAddr)
		if err != nil {
			ip = r.RemoteAddr
		}
		
		// Handle X-Forwarded-For if needed (Node.js version does)
		forwarded := r.Header.Get("X-Forwarded-For")
		if forwarded != "" {
			parts := strings.Split(forwarded, ",")
			ip = strings.TrimSpace(parts[0])
		}

		// Normalize IPv6 mapped IPv4
		if strings.HasPrefix(ip, "::ffff:") {
			ip = ip[7:]
		}

		if _, ok := whitelistMap[ip]; ok {
			next.ServeHTTP(w, r)
			return
		}

		if _, ok := blacklistMap[ip]; ok {
			log.Printf("Blocked request from blacklisted IP: %s", ip)
			http.Error(w, "Forbidden", http.StatusForbidden)
			return
		}

		if rateLimit.Enabled {
			f.mu.Lock()
			// Memory protection
			if len(f.requestCounts) > 20000 {
				f.requestCounts = make(map[string]*requestRecord)
				log.Println("Firewall request counts cleared due to memory limit")
			}

			now := time.Now().UnixMilli()
			record, exists := f.requestCounts[ip]

			if !exists {
				f.requestCounts[ip] = &requestRecord{count: 1, timestamp: now}
			} else {
				if now-record.timestamp > int64(rateLimit.WindowMs) {
					record.count = 1
					record.timestamp = now
				} else {
					record.count++
				}
			}

			count := f.requestCounts[ip].count
			f.mu.Unlock()

			if count > rateLimit.Max {
				if count == rateLimit.Max+1 {
					log.Printf("Rate limit exceeded for IP: %s", ip)
				}
				http.Error(w, "Too Many Requests", http.StatusTooManyRequests)
				return
			}
		}


		// WebSocket Connection Limit
		isWebSocket := strings.ToLower(r.Header.Get("Upgrade")) == "websocket"
		
		if isWebSocket && f.config.MaxWSPerIP > 0 {
			f.mu.Lock()
			currentWS := f.wsCounts[ip]
			
			if currentWS >= f.config.MaxWSPerIP {
				f.mu.Unlock()
				log.Printf("Blocked WebSocket from %s: Max concurrent connections (%d) reached", ip, f.config.MaxWSPerIP)
				http.Error(w, "Too Many WebSocket Connections", http.StatusTooManyRequests)
				return
			}
			
			f.wsCounts[ip]++
			f.mu.Unlock()
			
			// Decrement on completion
			defer func() {
				f.mu.Lock()
				f.wsCounts[ip]--
				if f.wsCounts[ip] <= 0 {
					delete(f.wsCounts, ip)
				}
				f.mu.Unlock()
			}()
		}

		next.ServeHTTP(w, r)
	})
}

func sliceToMap(slice []string) map[string]struct{} {
	m := make(map[string]struct{}, len(slice))
	for _, s := range slice {
		m[s] = struct{}{}
	}
	return m
}
